testing/async/test_policy_overloads.h

```diff
@@ -0,0 +1,421 @@
+    constexpr auto postfix_tuple_size = std::tuple_size<postfix_tuple_type>{};
+    using prefix_index_seq  = std::make_index_sequence<1>;
+    using postfix_index_seq = std::make_index_sequence<postfix_tuple_size>;
+
+    // Use unique, non-const inputs for each invocation to support in-place
+    // algo_def configurations.
+    input_type input_a   = algo_def::generate_input(num_values);
+    input_type input_b   = algo_def::generate_input(num_values);
+    input_type input_c   = algo_def::generate_input(num_values);
+    input_type input_tmp = algo_def::generate_input(num_values);
+    input_type input_ref = algo_def::generate_input(num_values);
+
+    output_type output_a   = algo_def::generate_output(num_values, input_a);
+    output_type output_b   = algo_def::generate_output(num_values, input_b);
+    output_type output_c   = algo_def::generate_output(num_values, input_c);
+    output_type output_tmp = algo_def::generate_output(num_values, input_tmp);
+    output_type output_ref = algo_def::generate_output(num_values, input_ref);
+
+    auto e_a = algo_def::invoke_async(std::make_tuple(thrust::device),
+                                      prefix_index_seq{},
+                                      input_a,
+                                      output_a,
+                                      postfix_tuple,
+                                      postfix_index_seq{});
+    ASSERT_EQUAL(true, e_a.valid_stream());
+    auto const stream_a = e_a.stream().native_handle();
+
+    // Execution on default stream should create a new stream in the result:
+    ASSERT_NOT_EQUAL_QUIET(thrust::cuda_cub::default_stream(), stream_a);
+
+    // Explicitly order this invocation after the previous one:
+    auto e_b =
+      algo_def::invoke_async(std::make_tuple(thrust::device.after(e_a)),
+                             prefix_index_seq{},
+                             input_b,
+                             output_b,
+                             postfix_tuple,
+                             postfix_index_seq{});
+    ASSERT_EQUAL(true, e_b.valid_stream());
+    auto const stream_b = e_b.stream().native_handle();
+
+    // Second invocation should use same stream as before:
+    ASSERT_EQUAL_QUIET(stream_a, stream_b);
+
+    // Verify that double consumption of e_a produces an exception:
+    ASSERT_THROWS_EQUAL(auto x = algo_def::invoke_async(
+                          std::make_tuple(thrust::device.after(e_a)),
+                          prefix_index_seq{},
+                          input_tmp,
+                          output_tmp,
+                          postfix_tuple,
+                          postfix_index_seq{});
+                        THRUST_UNUSED_VAR(x),
+                        thrust::event_error,
+                        thrust::event_error(thrust::event_errc::no_state));
+
+    // Explicitly order this invocation after e_b:
+    auto policy_after_e_b = thrust::device.after(e_b);
```

Comment:
We should leave a comment here explaining that it is important that in the first of these tests, the execution policy is an rvalue and in the second of these tests it is an lvalue; both are supposed to be consumable.

Reply:
Done.